src/crypto/tls/handshake_client_tls13.go

```diff
@@ -82,6 +82,44 @@ type clientHandshakeStateTLS13 struct {
 	handshakeTimings EXP_EventTLS13ClientHandshakeTimingInfo
 }
 
+// processDelegatedCredentialFromServer unmarshals the DelegatedCredential
+// offered by the server (if present) and validates it using the peer
+// certificate.
+func (hs *clientHandshakeStateTLS13) processDelegatedCredentialFromServer(dc []byte, certVerifyMsg *certificateVerifyMsg) error {
+	c := hs.c
+
+	var dCred *DelegatedCredential
+	var err error
+	if dc != nil {
+		// Assert that the DC extension was indicated by the client.
+		if !hs.hello.delegatedCredentialSupported {
+			c.sendAlert(alertIllegalParameter)
+			return errors.New("tls: got delegated credential extension without indication")
+		}
+
+		dCred, err = unmarshalDelegatedCredential(dc)
+		if err != nil {
+			c.sendAlert(alertDecodeError)
+			return fmt.Errorf("tls: delegated credential: %s", err)
+		}
+
+		if !isSupportedSignatureAlgorithm(dCred.cred.expCertVerfAlgo, supportedSignatureAlgorithmsDC) {
+			c.sendAlert(alertIllegalParameter)
+			return errors.New("tls: delegated credential used with invalid signature algorithm")
+		}
+	}
+
+	if dCred != nil {
+		if !dCred.Validate(c.peerCertificates[0], DCServer, c.config.time(), certVerifyMsg) {
```

Comment:
I think this was the subject of debate as well. In the sense of: if the validation of a dc fails, should we fail all-together or fall back to a certificate?

Reply:
My understanding is that fallback to the certificate isn't possible. Since the server has sent a DC at this point, we must presume that the DC was used to sign the handshake. Since the DC is invalid, we must hard-fail here to prevent a MITM attack. Note that this is also required by the spec: see https://tools.ietf.org/html/draft-ietf-tls-subcerts-10#section-4.1.3.

Reply:
I thought that too. I think I'll go with adding the alert back ;)

Reply:
Added the alert back ;)